File: metatoolkit/box.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import argparse
import matplotlib.pyplot as plt
import pandas as pd
from pathlib import Path
import seaborn as sns

def parse_arguments():
    parser = argparse.ArgumentParser(description='Produces a Boxplot of a given dataset')
    parser.add_argument('subject', help='Path to dataset file or subject name')
    parser.add_argument('-x', help='Column name for x-axis')
    parser.add_argument('-y', help='Column name for y-axis')
    parser.add_argument('--hue', help='Column name for hue grouping')
    parser.add_argument(
        '--order',
        help='Comma-separated order of categories for x-axis (or y-axis if horizontal)'
    )
    parser.add_argument(
        '--logy',
        action='store_true',
        help='Set y-axis to log scale (or x-axis if horizontal)'
    )
    parser.add_argument('--show', action='store_true', help='Display the plot window')
    parser.add_argument(
        '--figsize',
        default='2,2',
        help='Figure size per facet as width,height'
    )
    parser.add_argument('-o', '--output', help='Output filename without extension')
    parser.add_argument(
        '--meta',
        nargs='+',
        help='Path(s) to metadata file(s) to inner-join with subject data before plotting'
    )
    parser.add_argument(
        '--rc',
        help='Path to matplotlibrc file to use for styling'
    )
    parser.add_argument(
        '--horizontal',
        action='store_true',
        help='Plot horizontally (swap x and y axes)'
    )
    parser.add_argument('--ymin', type=float, help='Minimum value for y-axis')
    parser.add_argument('--ymax', type=float, help='Maximum value for y-axis')

    # Faceting arguments
    parser.add_argument('--row', help='Column name to facet across rows')
    parser.add_argument('--col', help='Column name to facet across columns')
    parser.add_argument(
        '--sharey',
        default='all',
        choices=['all', 'none', 'row', 'col'],
        help='Whether to share the y-axis across facets. Options: all (default), none, row, col'
    )

    # Color palette argument
    parser.add_argument(
        '--palette',
        default='pastel',
        help='Seaborn color palette for the boxplot (default: pastel)'
    )

    return parser.parse_args()

def load_data(path_or_name):
    path = Path(path_or_name)
    return pd.read_csv(path, sep='\t', index_col=0)

def merge_meta(df, meta_paths):
    for mpath in meta_paths:
        mdf = pd.read_csv(mpath, sep='\t', index_col=0)
        df = df.join(mdf, how='inner')
    return df

def plot_box(df, x, y, hue, figsize, horizontal=False, order=None, row=None, col=None, sharey=True, palette='pastel'):
    df = df.reset_index()

    # swap x/y if horizontal
    if horizontal:
        x, y = y, x

    order_list = order.split(",") if order else None

    x_col = x or df.columns[0]
    y_col = y or df.columns[1]

    # NEW: Lock the hue order globally so missing variables in subplots don't shift colors
    if hue:
        hue_order_list = df[hue].dropna().drop_duplicates().tolist()
    else:
        hue_order_list = None

    # Seaborn FacetGrid takes height (per facet) and aspect ratio (width/height)
    width, height = figsize
    aspect = width / height if height != 0 else 1

    g = sns.FacetGrid(
        data=df,
        row=row,
        col=col,
        sharey=sharey,
        height=height,
        aspect=aspect
    )

    g.map_dataframe(
        sns.boxplot,
        x=x_col,
        y=y_col,
        hue=hue,
        order=order_list,
        hue_order=hue_order_list,  # Enforce consistent colors
        palette=palette,
        showfliers=False,
        showcaps=False,
        linewidth=0.4,
        boxprops={'edgecolor': 'black'},
        whiskerprops={'color': 'black'},
        medianprops={'color': 'black'},
        capprops={'color': 'black'}
    )

    # Setup Stripplot arguments dynamically to avoid Seaborn Warnings
    strip_kwargs = {
        'x': x_col,
        'y': y_col,
        'order': order_list,
        'size': 1,
        'dodge': bool(hue)
    }

    if hue:
        strip_kwargs['hue'] = hue
        strip_kwargs['hue_order'] = hue_order_list  # Enforce consistent dodge positioning

        # Provide an array of black colors matching the exact number of global hues
        n_hues = len(hue_order_list)
        strip_kwargs['palette'] = ['black'] * n_hues

        # Suppress the stripplot legend so it doesn't overwrite the boxplot colors
        strip_kwargs['legend'] = False
    else:
        strip_kwargs['color'] = 'black'

    g.map_dataframe(
        sns.stripplot,
        **strip_kwargs
    )

    g.despine(right=True, top=True)

    if hue:
        g.add_legend()

    return g

def save_plots(filename, show):
    filename = Path(filename)
    if filename.suffix:  # If the filename has an extension
        out_path = filename
        out_path.parent.mkdir(parents=True, exist_ok=True)
    else:
        out_dir = Path('results')
        out_dir.mkdir(parents=True, exist_ok=True)
        out_path = out_dir / f'{filename}.svg'

    plt.savefig(out_path)
    if show:
        plt.show()
    plt.clf()

def main():
    args = parse_arguments()

    # load matplotlibrc file if specified
    if args.rc:
        plt.style.use(args.rc)

    # load subject data
    df = load_data(args.subject)

    # optionally merge in metadata explainers
    if args.meta:
        df = merge_meta(df, args.meta)

    # parse figsize
    figsize = tuple(map(float, args.figsize.split(',')))

    # Parse sharey argument mapping
    sharey_val = args.sharey
    if sharey_val == 'all':
        sharey_val = True
    elif sharey_val == 'none':
        sharey_val = False

    # plot
    g = plot_box(
        df,
        args.x,
        args.y,
        args.hue,
        figsize,
        horizontal=args.horizontal,
        order=args.order,
        row=args.row,
        col=args.col,
        sharey=sharey_val,
        palette=args.palette
    )

    # handle log scaling across all facet axes
    if args.logy:
        for ax in g.axes.flat:
            if args.horizontal:
                ax.set_xscale('log')
            else:
                ax.set_yscale('log')

    # apply y-axis (or x-axis) limits across the grid
    if args.horizontal:
        if args.ymin is not None or args.ymax is not None:
            g.set(xlim=(args.ymin, args.ymax))
    else:
        if args.ymin is not None or args.ymax is not None:
            g.set(ylim=(args.ymin, args.ymax))

    plt.tight_layout()

    # save (and optionally show)
    save_plots(args.output or f'{Path(args.subject).stem}_box', args.show)

if __name__ == '__main__':
    main()

File: metatoolkit/box.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(argparse)
  library(ggplot2)
  library(readr)
  library(dplyr)
  library(tibble)
  library(rlang)    # for sym()
})

# ---- Parse arguments ----
parser <- ArgumentParser(description = "Produces a Boxplot of a given dataset")
parser$add_argument("subject", help = "Path to dataset file or subject name")
parser$add_argument("-x", help = "Column name for x-axis")
parser$add_argument("-y", help = "Column name for y-axis")
parser$add_argument("--hue", help = "Column name for hue grouping")
parser$add_argument("--logy", action = "store_true", help = "Set y-axis to log scale")
parser$add_argument("--show", action = "store_true", help = "Display the plot window")
parser$add_argument("--figsize", default = "2,2", help = "Figure size as width,height (in inches)")
parser$add_argument("-o", "--output", help = "Output filename without extension")
parser$add_argument("--meta", nargs = "+", help = "Path(s) to metadata file(s) to inner-join")
parser$add_argument("--rc", help = "Path to ggplot2 theme file (RDS with theme)")

args <- parser$parse_args()

# ---- Load data ----
load_data <- function(path_or_name) {
  if (file.exists(path_or_name)) {
    df <- read_tsv(path_or_name)
  } else {
    fname <- file.path("results", paste0(path_or_name, ".tsv"))
    df <- read_tsv(fname)
  }
  df <- as.data.frame(df)
  rownames(df) <- df[[1]]
  df <- df[,-1, drop = FALSE]
  return(df)
}

# ---- Merge metadata ----
merge_meta <- function(df, meta_paths) {
  for (mpath in meta_paths) {
    mdf <- read_delim(mpath, delim = NULL, col_types = cols())
    mdf <- as.data.frame(mdf)
    rownames(mdf) <- mdf[[1]]
    mdf <- mdf[,-1, drop = FALSE]
    df <- inner_join(
      df %>% rownames_to_column("id"),
      mdf %>% rownames_to_column("id"),
      by = "id"
    ) %>% column_to_rownames("id")
  }
  return(df)
}

# ---- Plot function ----
plot_box <- function(df, x, y, hue, figsize) {
  df <- rownames_to_column(df, "index")

  x_col <- if (is.null(x)) colnames(df)[2] else x
  y_col <- if (is.null(y)) colnames(df)[3] else y

  # add small pseudocount to avoid log10(0)
  df[[y_col]][df[[y_col]] == 0] <- 1e-6

  # build base aesthetic
  if (is.null(hue)) {
    aes_mapping <- aes(x = !!sym(x_col), y = !!sym(y_col))
  } else {
    aes_mapping <- aes(x = !!sym(x_col), y = !!sym(y_col),
                       fill = !!sym(hue), color = !!sym(hue))
  }

  p <- ggplot(df, aes_mapping) +
    geom_boxplot(outlier.shape = NA, width = 0.6, alpha = 0.7) +
    geom_jitter(size = 1.2, width = 0.2, alpha = 0.8) +
    theme_classic(base_size = 10) +
    theme(
      panel.background = element_rect(fill = "#F5F0DC", colour = NA),
      axis.text  = element_text(color = "black"),
      axis.title = element_text(color = "black")
    )

  # Add colors only if hue is defined
  if (!is.null(hue)) {
    p <- p +
      scale_fill_manual(values = c("#C0DFA1", "#7FA960")) +
      scale_color_manual(values = c("#C0DFA1", "#7FA960"))
  }

  return(p)
}

# ---- Main ----
df <- load_data(args$subject)
print(df)

if (!is.null(args$meta)) {
  df <- merge_meta(df, args$meta)
}

# load theme if provided
if (!is.null(args$rc)) {
  theme_custom <- readRDS(args$rc)
  if (inherits(theme_custom, "theme")) {
    theme_set(theme_custom)
  }
}

# parse figsize
figsize <- as.numeric(strsplit(args$figsize, ",")[[1]])

# plot
p <- plot_box(df, args$x, args$y, args$hue, figsize)
if (args$logy) {
  p <- p + scale_y_log10()
}

# save
outfile <- if (!is.null(args$output)) {
  args$output
} else {
  file.path("results", paste0(tools::file_path_sans_ext(basename(args$subject)), "_box.svg"))
}
dir.create(dirname(outfile), showWarnings = FALSE, recursive = TRUE)

ggsave(outfile, plot = p, width = figsize[1], height = figsize[2])

if (args$show) {
  print(p)
}

